quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/services/metrics/events/MicrometerMetricsEventListener.java

```diff
@@ -0,0 +1,173 @@
+    private static final String OPERATION_TAG = "operation";
+    private static final String RESULT_TAG = "result";
+    private static final String SUCCESS = "success";
+    private static final String ERROR = "error";
+    private static final String EVENT_TYPE_ERROR_SUFFIX = "_ERROR";
+
+    private static final String KEYLOAK_METER_NAME_PREFIX = "keycloak.";
+    private static final String EVENT_PREFIX = KEYLOAK_METER_NAME_PREFIX + "event.";
+    private static final String ADMIN_EVENT_COUNTER_NAME = KEYLOAK_METER_NAME_PREFIX + "admin.event";
+
+    private static final Map<EventType, String> EVENT_TYPE_TO_NAME =
+            Arrays.stream(EventType.values())
+                    .collect(Collectors.toMap(e -> e, MicrometerMetricsEventListener::buildCounterName));
+
+    private final EventListenerTransaction tx =
+            new EventListenerTransaction(this::countRealmResourceTagsFromGenericAdminEvent, this::countEvent);
+
+    private final MeterRegistry meterRegistry = Metrics.globalRegistry;
+    private final EnumSet<EventType> includedEvents;
+    private final boolean adminEventEnabled;
+
+    public MicrometerMetricsEventListener(KeycloakSession session, EnumSet<EventType> includedEvents,
+                                          boolean adminEventEnabled) {
+        session.getTransactionManager().enlistAfterCompletion(tx);
+        this.includedEvents = includedEvents;
+        this.adminEventEnabled = adminEventEnabled;
+    }
+
+    @Override
+    public void onEvent(Event event) {
+        if (includedEvents.contains(event.getType())) {
+            tx.addEvent(event);
+        }
+    }
+
+    private void countEvent(Event event) {
+        logger.debugf("Received user event of type %s in realm %s",
+                event.getType().name(), event.getRealmName());
+        switch (event.getType()) {
+            case LOGIN:
+            case LOGIN_ERROR:
+            case CLIENT_LOGIN:
+            case CLIENT_LOGIN_ERROR:
+            case REGISTER:
+            case REGISTER_ERROR:
+            case REFRESH_TOKEN:
+            case REFRESH_TOKEN_ERROR:
+            case CODE_TO_TOKEN:
+            case CODE_TO_TOKEN_ERROR:
+                countRealmProviderClientIdResultErrorTagsFromEvent(event);
+                break;
+            default:
+                countRealmResultTagsFromEvent(event);
+        }
+    }
+
+    @Override
+    public void onEvent(AdminEvent event, boolean includeRepresentation) {
+        if (adminEventEnabled) {
+            tx.addAdminEvent(event, includeRepresentation);
+        }
+    }
+
+    private void countRealmResultTagsFromEvent(final Event event) {
+        meterRegistry.counter(EVENT_TYPE_TO_NAME.get(event.getType()),
+                REALM_TAG, nullToEmpty(event.getRealmName()),
+                RESULT_TAG, getResultCode(event)).increment();
+    }
+
+    private void countRealmProviderClientIdResultErrorTagsFromEvent(final Event event) {
```

Comment:
I'm not sure that we should include the provider tag. Maybe it should be configurable?